mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,662 @@
+            RecQNetwork: Recurrent Q network
+            optax.GradientTransformation: Optimiser for RecQNetwork.
+            TrajectoryBuffer: The replay buffer.
+            LearnerState: The initial learner state.
+            MavaLogger: The logger.
+            PRNGKey: The random key.
+    """
+    logger = MavaLogger(cfg)
+
+    key = jax.random.PRNGKey(cfg.system.seed)
+    devices = jax.devices()
+
+    def replicate(x: Any) -> Any:
+        """First replicate the update batch dim then put on devices."""
+        x = jax.tree_map(lambda y: jnp.broadcast_to(y, (cfg.system.update_batch_size, *y.shape)), x)
+        return jax.device_put_replicated(x, devices)
+
+    env, eval_env = environments.make(cfg)
+
+    action_dim = env.action_dim
+    num_agents = env.num_agents
+
+    key, q_key = jax.random.split(key, 2)
+    # Shape legend:
+    # T: Time
+    # B: Batch
+    # A: Agent
+    # Make dummy inputs to init recurrent Q network -> need shape (T, B, A, ...)
+    init_obs = env.observation_spec().generate_value()  # (A, ...)
+    init_obs_batched = jax.tree_util.tree_map(
+        lambda x: jnp.repeat(x[jnp.newaxis, ...], cfg.arch.num_envs, axis=0),
+        init_obs,
+    )  # (B, A, ...)
+    init_obs_batched = jax.tree_util.tree_map(
+        lambda x: x[jnp.newaxis, ...], init_obs_batched
+    )  # add time dim (1, B, A, ...)
+    init_done = jnp.zeros((1, cfg.arch.num_envs, 1), dtype=bool)  # (1, B, 1)
+    init_x = (init_obs_batched, init_done)  # pack the RNN dummy inputs
+    init_hidden_state = ScannedRNN.initialize_carry(
+        (cfg.arch.num_envs, num_agents), cfg.network.hidden_state_dim
+    )  # (B, A, ...)
+
+    # Making recurrent Q network
+    pre_torso = hydra.utils.instantiate(cfg.network.q_network.pre_torso)
+    post_torso = hydra.utils.instantiate(cfg.network.q_network.post_torso)
+    q_net = RecQNetwork(
+        pre_torso,
+        post_torso,
+        action_dim,
+        cfg.network.hidden_state_dim,
+    )
+    q_params = q_net.init(q_key, init_hidden_state, init_x)  # epsilon defaults to 0
+    q_target_params = q_net.init(q_key, init_hidden_state, init_x)  # ensure parameters are separate
+
+    # Pack Q network params
+    params = DDQNParams(q_params, q_target_params)
+
+    # Making optimiser and state
+    if cfg.system.clip_optimiser:
+        opt = optax.chain(
+            optax.clip_by_global_norm(cfg.system.global_norm),
+            optax.adam(learning_rate=cfg.system.q_lr, eps=1e-5),
+        )
+    else:
+        opt = optax.adam(learning_rate=cfg.system.q_lr)
+    opt_state = opt.init(params.online)
+
+    # Distribute params, opt states and hidden states across all devices
+    params = replicate(params)
+    opt_state = replicate(opt_state)
+    init_hidden_state = replicate(init_hidden_state)
+
+    # Create dummy transition
+    init_acts = env.action_spec().generate_value()  # (A,)
+    init_transition = Transition(
+        obs=Observation(*init_obs),  # (A, ...)
```

Comment:
Why do you need to wrap this in an observation? `init_obs` should already be this type?
```suggestion
        obs=init_obs,  # (A, ...)
```

Reply:
Good catch! Just checking that it doesn't break anything, then will commit.